### ISSUE TITLE
unable to restart after crash timeout interruption etc

### ISSUE THREAD

**Issue by 1HOD** [BitMEX/sample-market-maker#188]
After any exception that halts the MM I get this:
-----
2019-03-25 20:32:31,746 - INFO - market_maker - Restarting the market maker...

C:\Users\Vostro 3700>c:\users\vostro: can't open file '3700\appdata\local\progra
ms\python\python37-32\python.exe': [Errno 2] No such file or directory

I presume this is because the script tells it to look for python.exe and can't find it in that location.  the restart function in market_maker.py seems to read like this:

    def restart(self):
        logger.info("Restarting the market maker...")
        os.execv(sys.executable, [sys.executable] + sys.argv)

execv seems to require specific information in the right order to function correctly, I'm guessing it's as simple as I don't have the right file in the right folder rather than a coding error.  Could someone point me in the right direction please? I'd like to try and get the 'automatic restart' functioning as its crippled the bot many times previously.

Many thanks!

**Comment by neurotronix** [comment]
If you have access to a bash/zsh shell, then you can simply use a shell launch script. This will run the python script until it finishes (returns 0) or restarts when it crashes (returns 1).
```shell
#!/usr/bin/env bash

MM="python ./marketmaker"

until $MM; do
    echo "$MM crashed with exit code $?. Restarting ..." >&2
done
```
I use this script in combination with nohup to run the MM in detached mode: `nohup sh ./launch.sh &`

**Comment by hnasir1** [comment]
> If you have access to a bash/zsh shell, then you can simply use a shell launch script. This will run the python script until it finishes (returns 0) or restarts when it crashes (returns 1).
> 
> ```shell
> #!/usr/bin/env bash
> 
> MM="python ./marketmaker"
> 
> until $MM; do
>     echo "$MM crashed with exit code $?. Restarting ..." >&2
> done
> ```
> 
> I use this script in combination with nohup to run the MM in detached mode: `nohup sh ./launch.sh &`

So even if system overload happens and it disconnects the MM, it will automatically restart and continue where it left off?

**Comment by neurotronix** [comment]
Yes, if the MM crashes this script will restart the whole thing. But be aware of open orders after the crash. When MM restarts, it places new orders. So you might end up with a lot (unintended) open orders and exposure if they get filled.
